File: android/src/main/java/expo/modules/applinks/ExpoApplinksModule.kt
package expo.modules.applinks

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.Exceptions
import com.applinks.android.AppLinksSDK
import com.applinks.android.AppLinksSDKVersion
import android.net.Uri
import android.os.Handler
import android.os.Looper
import com.applinks.android.LinkType
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import androidx.core.net.toUri
import com.applinks.android.middleware.LinkHandlingResult

class ExpoApplinksModule : Module() {
  private var linkListener: AppLinksSDK.AppLinksListener? = null
  private val mainHandler = Handler(Looper.getMainLooper())
  
  override fun definition() = ModuleDefinition {
    Name("ExpoApplinks")

    Events("onLinkHandled")
    
    OnStartObserving {
      setupLinkListener()
    }
    
    OnStopObserving {
      removeLinkListener()
    }

    AsyncFunction("initialize") { config: Map<String, Any>, promise: Promise ->
      try {
        val apiKey = config["apiKey"] as? String ?: throw Exception("API key is required")
        val autoHandleLinks = config["autoHandleLinks"] as? Boolean ?: false
        
        val logLevel = when (config["logLevel"] as? String) {
          "none" -> "none"
          "error" -> "error"
          "warning" -> "warning"
          "info" -> "info"
          "debug" -> "debug"
          else -> "info"
        }
        
        val context = appContext.reactContext ?: throw Exceptions.AppContextLost()
        
        // Get plugin configuration from manifest
        val metadata = context.packageManager.getApplicationInfo(
          context.packageName,
          android.content.pm.PackageManager.GET_META_DATA
        ).metaData
        
        val supportedDomains = metadata?.getString("com.applinks.supportedDomains")?.split(",") ?: emptyList()
        val supportedSchemes = metadata?.getString("com.applinks.supportedSchemes")?.split(",") ?: emptyList()
        
        // Initialize AppLinksSDK
        val sdk = AppLinksSDK.builder(context)
          .apiKey(apiKey)
          .apply {
            if (supportedDomains.isNotEmpty()) {
              supportedDomains(*supportedDomains.toTypedArray())
            }
            if (supportedSchemes.isNotEmpty()) {
              supportedSchemes(*supportedSchemes.toTypedArray())
            }
          }
          .build()
        
        // Mark SDK as initialized and process any pending URLs
        ExpoApplinksPlugin.markSDKInitialized(autoHandleLinks)
        
        promise.resolve(null)
      } catch (e: Exception) {
        promise.reject("InitializationError", e.message, e)
      }
    }

    Function("getVersion") {
      return@Function AppLinksSDKVersion.current
    }

    AsyncFunction("getAppLinkDetails") { uri: String, promise: Promise ->
      CoroutineScope(Dispatchers.IO).launch {
        try {
          val handledLink = AppLinksSDK.getInstance().getAppLinkDetails(uri.toUri())
          if (!handledLink.handled) {
            promise.resolve(null)
            return@launch
          }

          val result = mapOf(
            "handled" to handledLink.handled,
            "originalUrl" to handledLink.originalUrl.toString(),
            "path" to handledLink.path,
            "params" to handledLink.params,
            "metadata" to handledLink.metadata,
            "error" to handledLink.error
          )

          promise.resolve(result)
        } catch (e: Exception) {
          promise.resolve(e.message ?: "Failed to retrieve link")
        }
      }
    }

    AsyncFunction("getInitialLink") { promise: Promise ->
      val initialLink = ExpoApplinksPlugin.initialLink
      if (initialLink == null) {
        promise.resolve(null)
        return@AsyncFunction
      }
      
      // Launch coroutine to call suspend function
      CoroutineScope(Dispatchers.IO).launch {
        try {
          val handledLink = AppLinksSDK.getInstance().getAppLinkDetails(initialLink)
          if (!handledLink.handled) {
            promise.resolve(null)
            return@launch
          }
          
          val result = mapOf(
            "handled" to handledLink.handled,
            "originalUrl" to handledLink.originalUrl.toString(),
            "path" to handledLink.path,
            "params" to handledLink.params,
            "metadata" to handledLink.metadata,
            "error" to handledLink.error
          )

          promise.resolve(result)
        } catch (e: Exception) {
          promise.resolve(null)
        }
      }
    }

    AsyncFunction("createLink") { params: Map<String, Any>, promise: Promise ->
      try {
        val domain = params["domain"] as? String ?: throw Exception("Domain is required")
        val type = params["type"] as? String ?: "short"
        val title = params["title"] as? String ?: throw Exception("Title is required")
        val subtitle = params["subtitle"] as? String
        val deepLinkPath = params["deepLinkPath"] as? String ?: throw Exception("Deep link path is required")
        val deepLinkParams = params["deepLinkParams"] as? Map<String, String> ?: emptyMap()
        val webLink = params["web_link"] as? String
        val expiresAt = params["expiresAt"] as? Long // Unix timestamp in milliseconds
        val backgroundType = params["background_type"] as? String
        val backgroundColor = params["background_color"] as? String
        val backgroundColorFrom = params["background_color_from"] as? String
        val backgroundColorTo = params["background_color_to"] as? String
        val backgroundColorDirection = params["background_color_direction"] as? String
        
        AppLinksSDK.getInstance().linkShortener.createLinkAsync {
          this.webLink = webLink?.let { Uri.parse(it) }
          this.domain = domain
          this.title = title
          this.subtitle = subtitle
          this.deepLinkPath = deepLinkPath
          this.deepLinkParams = deepLinkParams
          this.backgroundType = backgroundType
          this.backgroundColor = backgroundColor
          this.backgroundColorFrom = backgroundColorFrom
          this.backgroundColorTo = backgroundColorTo
          this.backgroundColorDirection = backgroundColorDirection
          linkType = when (type) {
            "unguessable" -> LinkType.UNGUESSABLE
            "short" -> LinkType.SHORT
            else -> LinkType.SHORT
          }
          expiresAt?.let {
            this.expiresAt = it
          }
        }.addOnSuccessListener { result ->
          val (shortLink) = result
          promise.resolve(shortLink)
        }.addOnFailureListener { exception ->
          promise.reject("CreateLinkError", exception.message ?: "Failed to create link", exception)
        }
      } catch (e: Exception) {
        promise.reject("CreateLinkError", e.message, e)
      }
    }
  }
  
  private fun setupLinkListener() {
    if (linkListener == null) {
      linkListener = object : AppLinksSDK.AppLinksListener {
        override fun onLinkReceived(result: LinkHandlingResult) {
          sendLinkResult(result)
        }
        
        override fun onError(error: String) {
          val errorResult = mapOf(
            "handled" to false,
            "originalUrl" to "",
            "path" to "",
            "params" to emptyMap<String, String>(),
            "metadata" to emptyMap<String, Any>(),
            "error" to error
          )
          mainHandler.post {
            sendEvent("onLinkHandled", errorResult)
          }
        }
      }
      
      try {
        AppLinksSDK.getInstance().addLinkListener(linkListener!!)
      } catch (e: Exception) {
        // SDK not initialized yet
      }
    }
  }
  
  private fun removeLinkListener() {
    linkListener?.let {
      try {
        AppLinksSDK.getInstance().removeLinkListener(it)
      } catch (e: Exception) {
        // SDK not initialized
      }
      linkListener = null
    }
  }
  
  private fun sendLinkResult(result: LinkHandlingResult) {
    val eventData = mapOf(
      "handled" to result.handled,
      "originalUrl" to result.originalUrl.toString(),
      "path" to result.path,
      "params" to result.params,
      "metadata" to result.metadata,
      "error" to result.error
    )
    
    mainHandler.post {
      sendEvent("onLinkHandled", eventData)
    }
  }
}

File: ios/ExpoApplinksModule.swift
import ExpoModulesCore
import AppLinksSDK
import Combine

public class ExpoApplinksModule: Module {
  private var linkSubscription: AnyCancellable?
    
  public func definition() -> ModuleDefinition {
    Name("ExpoApplinks")

    Events("onLinkHandled")
    
    OnStartObserving {
      self.linkSubscription = AppLinksSDK.shared.linkPublisher
        .sink { [weak self] linkResult in
          self?.sendLinkResult(linkResult)
        }
    }
    
    OnStopObserving {
      self.linkSubscription?.cancel()
    }

    AsyncFunction("initialize") { (config: [String: Any]) in
      guard let apiKey = config["apiKey"] as? String else {
        throw Exception(name: "InvalidConfig", description: "API key is required")
      }
      
      let logLevel: AppLinksSDKLogLevel
      if let logLevelString = config["logLevel"] as? String {
        switch logLevelString {
        case "none": logLevel = .none
        case "error": logLevel = .error
        case "warning": logLevel = .warning
        case "info": logLevel = .info
        case "debug": logLevel = .debug
        default: logLevel = .info
        }
      } else {
        logLevel = .info
      }
      
      // Get plugin configuration from Info.plist
      let supportedDomains = Set(Bundle.main.object(forInfoDictionaryKey: "ExpoAppLinksSupportedDomains") as? [String] ?? [])
      let supportedSchemes = Set(Bundle.main.object(forInfoDictionaryKey: "ExpoAppLinksSupportedSchemes") as? [String] ?? [])
      
      // Initialize AppLinksSDK
      AppLinksSDK.initialize(
        apiKey: apiKey,
        supportedDomains: supportedDomains,
        supportedSchemes: supportedSchemes,
        logLevel: logLevel
      )
      
      // Mark SDK as initialized and process any pending URLs
      ExpoApplinksAppDelegate.markSDKInitialized()
    }

    AsyncFunction("createLink") { (params: [String: Any]) -> String in
      guard let domain = params["domain"] as? String,
            let type = params["type"] as? String,
            let title = params["title"] as? String,
            let deepLinkPath = params["deepLinkPath"] as? String else {
        throw Exception(name: "InvalidParams", description: "Required parameters are missing")
      }
      
      let linkType: LinkType
      switch type {
      case "unguessable":
        linkType = .unguessable
      case "short":
        linkType = .short
      default:
        throw Exception(name: "InvalidParams", description: "Invalid link type")
      }
      
      let subtitle = params["subtitle"] as? String
      let deepLinkParams = params["deepLinkParams"] as? Dictionary<String, String> ?? [:]
      let webLink = params["web_link"] as? String
      let backgroundType = params["background_type"] as? String
      let backgroundColor = params["background_color"] as? String
      let backgroundColorFrom = params["background_color_from"] as? String
      let backgroundColorTo = params["background_color_to"] as? String
      let backgroundColorDirection = params["background_color_direction"] as? String
      
      var expirationDate: Date? = nil
      if let expirationTimestamp = params["expiration_date"] as? Double {
        expirationDate = Date(timeIntervalSince1970: expirationTimestamp / 1000)
      }
      
      return try await AppLinksSDK.shared.linkShortener.createLink(
        LinkCreationParams(
            domain: domain,
            title: title,
            subtitle: subtitle,
            deepLinkPath: deepLinkPath,
            webLink: webLink,
            deepLinkParams: deepLinkParams,
            expiresAt: expirationDate,
            linkType: linkType,
            backgroundType: backgroundType,
            backgroundColor: backgroundColor,
            backgroundColorFrom: backgroundColorFrom,
            backgroundColorTo: backgroundColorTo,
            backgroundColorDirection: backgroundColorDirection
        )
      ).fullUrl
    }

    Function("getVersion") {
      return AppLinksSDK.version
    }
    
    AsyncFunction("getAppLinkDetails") { (url: String) -> [String: Any]? in
      guard let url = URL(string: url) else {
        return nil
      }
      
      let handledLink = await AppLinksSDK.shared.getAppLinkDetails(from: url)
      return [
        "handled": handledLink.handled,
        "originalUrl": handledLink.originalUrl.absoluteString,
        "path": handledLink.path,
        "params": handledLink.params,
        "metadata": handledLink.metadata,
        "error": handledLink.error as Any
      ]
    }
    
    AsyncFunction("getInitialLink") { () -> [String: Any]? in
      guard let link = await ExpoApplinksAppDelegate.initialLink else { return nil }
      
      let handledLink = await AppLinksSDK.shared.getAppLinkDetails(from: link)
      if (!handledLink.handled) {
        return nil
      }
        
      return [
        "handled": handledLink.handled,
        "originalUrl": handledLink.originalUrl.absoluteString,
        "path": handledLink.path,
        "params": handledLink.params,
        "metadata": handledLink.metadata,
        "error": handledLink.error as Any
      ]
    }
  }
  
  private func sendLinkResult(_ result: LinkHandlingResult) {
    let eventData: [String: Any] = [
      "handled": result.handled,
      "originalUrl": result.originalUrl.absoluteString,
      "path": result.path,
      "params": result.params,
      "metadata": result.metadata,
      "error": result.error as Any
    ]
    
    sendEvent("onLinkHandled", eventData)
  }
}
